rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md

```diff
@@ -0,0 +1,456 @@
+# RFC 192 - `minCapacity` ensures workers do not get unnecessarily killed
+* Comments: [#192](https://github.com/taskcluster/taskcluster-rfcs/pull/192)
+* Proposed by: @johanlorenzo
+
+# Summary
+
+Optimize worker pools with `minCapacity >= 1` by implementing minimum capacity workers that avoid unnecessary shutdown/restart cycles, preserving caches and reducing task wait times.
+
+## Motivation
+
+Currently, workers in pools with `minCapacity >= 1` exhibit wasteful behavior:
+
+1. **Cache Loss**: Workers shut down after idle timeout (600 seconds for decision pools), losing valuable caches:
+   - VCS repositories and history
+   - Package manager caches (npm, pip, cargo, etc.)
+   - Container images and layers
+
+2. **Provisioning Delays**: New worker provisioning takes ~73-74 seconds average for decision pools, during which tasks must wait
+
+3. **Resource Waste**: The current cycle of shutdown → detection → spawn → provision → register wastes compute resources and increases task latency
+
+4. **Violation of MinCapacity Intent**: `minCapacity >= 1` suggests these pools should always have capacity available, but the current implementation allows temporary capacity gaps
+
+# Details
+
+## Current Behavior Analysis
+
+**Affected Worker Pools:**
+- Direct `minCapacity: 1`: [`infra/build-decision`](https://github.com/mozilla-releng/fxci-config/blob/43c18aab0826244e369b16a964637b6c411c7760/worker-pools.yml#L220), [`code-review/bot-gcp`](https://github.com/mozilla-releng/fxci-config/blob/43c18aab0826244e369b16a964637b6c411c7760/worker-pools.yml#L3320)
+- Keyed `minCapacity: 1`: [`gecko-1/decision-gcp`, `gecko-3/decision-gcp`](https://github.com/mozilla-releng/fxci-config/blob/43c18aab0826244e369b16a964637b6c411c7760/worker-pools.yml#L976), and [pools matching `(app-services|glean|mozillavpn|mobile|mozilla|translations)-1`](https://github.com/mozilla-releng/fxci-config/blob/43c18aab0826244e369b16a964637b6c411c7760/worker-pools.yml#L1088)
+
+**Current Implementation Issues:**
+- Worker-manager enforces minCapacity by spawning new workers when capacity drops below threshold
+- Generic-worker shuts down after `afterIdleSeconds` regardless of minCapacity requirements
+- Gap exists between worker shutdown and replacement detection/provisioning
+
+## Proposed Solution
+
+### Core Concept: MinCapacity Workers
+
+Workers fulfilling `minCapacity >= 1` requirements should:
+1. **Extended Idle Timeout**: Receive significantly longer idle timeouts to preserve caches
+2. **Priority Replacement**: Be immediately replaced when they fail or become unhealthy
```

Comment:
challenge here would be to let worker manager track which workers were "promoted" to the long-living ones. And every time the worker is getting removed/killed, worker manager would need to scan all workers to know how many long-lived ones are still running .
If it would be tracked separately, then it would likely change the estimator and provisioner to treat them in a certain way.

I'm a bit concerned if this special treatment (immediate replacement)  is the right approach.

I think a better approach might be to not immediately replace it, but let all tasks drain, and as soon as there will be `totalRunning < minCapacity` - worker runner could provision this difference with the extended timeout workers. But not during `totalRunning >>> minCapacity` periods.

Reply:
Good call! I updated the RFC to highlight we just want to provision such workers when `totalRunning < minCapacity`.